siriuspy/siriuspy/idff/main.py

```diff
@@ -0,0 +1,226 @@
+
+
+class App(_Callback):
+    """Main application for handling machine shift."""
+
+    def __init__(self, idname):
+        """Class constructor."""
+        super().__init__()
+        self._const = _Const(idname)
+        self._pvs_prefix = self._const.idffname
+        self._pvs_database = self._const.get_propty_database()
+
+        self._loop_state = _Const.DsblEnbl.Dsbl
+        self._loop_freq = _Const.DEFAULT_LOOP_FREQ
+        self._control_qs = _Const.DsblEnbl.Dsbl
+        self._polarization = 'none'
+        self._config_name = ''
+        self.read_autosave_file()
+
+        # connections
+        # use pyepics recommendations for threading
+        _epics.ca.use_initial_context()
+
+        self._idff = _IDFF(idname)
+
+        # pvs to write methods
+        self.map_pv2write = {
+            'LoopState-Sel': self.set_loop_state,
+            'LoopFreq-SP': self.set_loop_freq,
+            'ControlQS-Sel': self.set_control_qs,
+            'ConfigName-SP': self.set_config_name,
+        }
+
+        self.quit = False
+        self._thread_ff = _epics.ca.CAThread(
+            target=self._do_ff, daemon=True)
+        self._thread_ff.start()
+
+    def init_database(self):
+        """Set initial PV values."""
+        pvn2vals = {
+            'LoopState-Sel': self._loop_state,
+            'LoopState-Sts': self._loop_state,
+            'LoopFreq-SP': self._loop_freq,
+            'LoopFreq-RB': self._loop_freq,
+            'ControlQS-Sel': self._control_qs,
+            'ControlQS-Sts': self._control_qs,
+            'ConfigName-SP': self._config_name,
+            'ConfigName-RB': self._config_name,
+            'Polarization-Mon': self._polarization,
+        }
+        for pvn, val in pvn2vals.items():
+            self.run_callbacks(pvn, val)
+        self._load_config(self._config_name)
+        self._update_log('Started.')
+
+    @property
+    def pvs_prefix(self):
+        """Return pvs_prefix."""
+        return self._pvs_prefix
+
+    @property
+    def pvs_database(self):
+        """Return pvs_database."""
+        return self._pvs_database
+
+    def process(self, interval):
+        """Sleep."""
+        _time.sleep(interval)
+
+    def read(self, reason):
+        """Read from IOC database."""
+        value = None
+        return value
+
+    def write(self, reason, value):
+        """Write value to reason and let callback update PV database."""
+        _log.info('Write received for: %s --> %s', reason, str(value))
+        if reason in self.map_pv2write.keys():
+            status = self.map_pv2write[reason](value)
+            _log.info('%s Write for: %s --> %s',
+                      str(status).upper(), reason, str(value))
+            return status
+        _log.warning('PV %s does not have a set function.', reason)
+        return False
+
+    def set_loop_state(self, value):
+        """Set loop state."""
+        if not 0 <= value < len(_ETypes.OPEN_CLOSED):
```

Comment:
does pcaspy not take care already of this checking? I dont recall.
if not, I think we should also update `Log-Mon` indicating the error.

Reply:
for enum PVs pcaspy do not take care of the checking.
I don't think it is necessary for this case to log the error, but I will do.